activerecord/lib/active_record/relation/query_methods.rb

```diff
@@ -303,6 +303,151 @@ def _select!(*fields) # :nodoc:
       self
     end
 
+    # Add a Common Table Expression (CTE) that you can then reference within another SELECT statement.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0))
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts
+    #
+    # Once you define Common Table Expression you can use custom `FROM` value or `JOIN` to reference it.
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).from("posts_with_tags AS posts")
+    #   # WITH posts_with_tags AS (
+    #   #  SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts_with_tags AS posts
+    #
+    #   Post.with(:posts_with_tags, Post.where("tags_count > ?", 0).joins("JOIN posts_with_tags ON posts_with_tags.id = posts.id"))
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE (tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts JOIN posts_with_tags ON posts_with_tags.id = posts.id
+    #
+    # You can also pass the expression as a `String`.
+    #
+    #   Post.with(:posts_with_tags, "SELECT * FROM posts WHERE tags_count > 0")
+    #   # WITH posts_with_tags AS (
+    #   #   SELECT * FROM posts WHERE tags_count > 0)
+    #   # )
+    #   # SELECT * FROM posts
+    #
+    # Note that building your own string from user input may expose your application to injection attacks
+    # if not done properly. It is recommended to use `ActiveRecord::Relation` instead.
+    #
+    # To add multiple CTEs you can chain `.with` / `.with_recursive` calls
+    #
+    #   Post
+    #     .with(:posts_with_comments, Post.where("comments_count > ?", 0))
+    #     .with(:posts_with_tags, Post.where("tags_count > ?", 0))
```

Comment:
Same here. Having both `.with(:posts_with_tags, Post.where("tags_count > ?", 0))` and `Post.with(posts_with_comments: Post.where("comments_count > ?", 0))` doesn't seem very valuable but adds lots of complexity.

I think it would be better to focus on a single way to do CTE.

Reply:
:+1: IMHO it will be nice to support hash-like interface and be able to specify multiple CTEs in one `with` call.

```ruby
Post.with(
  posts_with_comments: Post.where("comments_count > ?", 0),
  posts_without_comments: Post.where("comments_count = ?", 0)
)
```

Reply:
Yes, that makes most sense.

Reply:
We support only hash-like interface now.